tests/test_examples.py

```diff
@@ -0,0 +1,136 @@
+"""
+Test that the examples run without error.
+"""
+
+import os
+from pathlib import Path
+import importlib
+from unittest.mock import patch
+
+import imageio
+import numpy as np
+
+import pytest
+
+
+examples_dir = Path(__file__).parent.parent / "examples"
+screenshots_dir = examples_dir / "screenshots"
+diffs_dir = screenshots_dir / "diffs"
+
+# find examples that contain the marker comment for inclusion in the test suite
+MARKER_COMMENT = "# test_example = true"
+examples_to_test = []
+for example_path in examples_dir.glob("*.py"):
+    example_code = example_path.read_text()
+    if MARKER_COMMENT in example_code:
+        examples_to_test.append(example_path.stem)
+
+
+@pytest.fixture(autouse=True, scope="module")
+def force_offscreen():
+    """Force the offscreen canvas to be selected by the auto gui module."""
+    os.environ["WGPU_FORCE_OFFSCREEN"] = "true"
+    try:
+        yield
+    finally:
+        del os.environ["WGPU_FORCE_OFFSCREEN"]
+
+
+@pytest.fixture(autouse=True)
+def mock_time():
+    """Some examples use time to animate. Fix the return value
+    for repeatable output."""
+    with patch("time.time") as time_mock:
+        time_mock.return_value = 1.23456
+        yield
+
+
+@pytest.mark.parametrize("module", examples_to_test)
+def test_examples(module, pytestconfig):
+    """Run every example marked for testing."""
+    example = importlib.import_module(f"examples.{module}")
+
+    # render
+    img = example.canvas.draw()
+
+    # assert something was rendered
+    assert img is not None and img.size > 0
+
+    # regenerate screenshot if requested
+    screenshot_path = screenshots_dir / f"{module}.png"
+    if pytestconfig.getoption("regenerate_screenshots"):
+        imageio.imwrite(screenshot_path, img)
+
+    # if a reference screenshot exists, assert it is equal
+    if screenshot_path.exists():
+        stored_img = imageio.imread(screenshot_path)
+        # assert the basics; shape and dtype must match
+        assert img.dtype == stored_img.dtype
+        assert img.shape == stored_img.shape
+        # assert similarity
+        is_similar = images_similar(stored_img, img, atol=1)
+        if not is_similar:
+            # if the test fails, write the new image
+            diff_path = diffs_dir / f"{module}.png"
+            imageio.imwrite(diff_path, img)
+            # and a diff image, so it can be inspected
+            diff_path = diffs_dir / f"{module}-diff.png"
+            diff_img = np.abs(stored_img - img)
+            imageio.imwrite(diff_path, diff_img)
```

Comment:
Should we also remove old diff images that now pass?

Reply:
If a test fails, it will write the new image and the diff image to the `screenshots/diffs` subfolder.

That way you can look at the original in the `screenshots` folder, and the new and the diff images in the `diffs` subfolder.

Is that not what you want?

Reply:
Oh, now I get it. You mean, from previous test runs. You could add that, yeah.

Reply:
nah